File: video-service/src/client/java/com/boclips/videos/service/client/internal/resources/CollectionResource.java
package com.boclips.videos.service.client.internal.resources;

import com.boclips.videos.service.client.*;
import com.fasterxml.jackson.annotation.JsonProperty;
import lombok.Data;

import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

@Data
public class CollectionResource {
    private CollectionLinks _links;
    private String title;
    private String description;
    private Set<SubjectResource> subjects;
    private List<VideoResource> videos;
    @JsonProperty("public")
    private Boolean isPublic;
    private Boolean mine;

    public Collection toCollection() {
        Set<Subject> subjects = this.subjects.stream()
                .map(subjectResource -> Subject.builder()
                        .name(subjectResource.getName())
                        .id(SubjectId.builder().value(subjectResource.getId()).build())
                        .build())
                .collect(Collectors.toSet());

        List<Video> videos = this.videos.stream()
                .map(VideoResource::toVideo)
                .collect(Collectors.toList());

        return Collection.builder()
                .collectionId(new CollectionId(_links.getSelf().toUri()))
                .title(title)
                .description(description)
                .videos(videos)
                .isPublic(isPublic)
                .mine(mine)
                .subjects(subjects)
                .build();
    }
}

File: video-service-api/src/main/kotlin/com/boclips/videos/api/response/collection/CollectionResource.kt
package com.boclips.videos.api.response.collection

import com.boclips.videos.api.response.HateoasLink
import com.boclips.videos.api.response.agerange.AgeRangeResource
import com.boclips.videos.api.response.subject.SubjectResource
import com.boclips.videos.api.response.video.VideoResource
import com.fasterxml.jackson.annotation.JsonIgnore
import com.fasterxml.jackson.annotation.JsonInclude
import java.time.ZonedDateTime

data class CollectionResource(
    val id: String? = null,
    val owner: String? = null,
    val title: String? = null,
    val videos: List<VideoResource>,
    val updatedAt: ZonedDateTime? = null,
    val discoverable: Boolean? = null,
    val promoted: Boolean? = null,
    @get:JsonIgnore
    val bookmarked: Boolean? = null,
    val mine: Boolean? = null,
    val createdBy: String? = null,
    val subjects: Set<SubjectResource>,
    val ageRange: AgeRangeResource?,
    val description: String? = null,
    val attachments: Set<AttachmentResource>?,

    @JsonInclude(JsonInclude.Include.NON_NULL)
    var _links: Map<String, HateoasLink>? = null
)
